### PR TITLE
topology2: add src support

### PR DESCRIPTION
**Pull request by RanderWang**
This PR add ipc src support in topology2.

### CODE REVIEW

File: tools/topology/topology2/include/components/src-cavs.conf [hunk 1]
```diff
@@ -0,0 +1,96 @@
+#
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/components/src-cavs.conf:1]: there is a src.conf for ipc3

[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/components/src-cavs.conf]: refine it. Thanks

File: tools/topology/topology2/include/common/audio_format.conf [hunk 2]
```diff
@@ -232,6 +232,6 @@ Class.Base."audio_format" {
 	out_fmt_cfg	"$[($out_channels | ($out_valid_bit_depth * 256)) | ($out_sample_type * 65536)]"
 
 	# math expression for computing input/put buffer sizes
-	ibs "$[($in_channels * ($in_rate / 1000)) * ($in_bit_depth / 8)]"
-	obs "$[($out_channels * ($out_rate / 1000)) * ($out_bit_depth / 8)]"
+	ibs "$[($in_channels * ($[($in_rate + 900)] / 1000)) * ($in_bit_depth / 8)]"
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/common/audio_format.conf]: sure, thanks

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+	direction	"playback"
+	dynamic_pipeline 1
+	time_domain	"timer"
+	channels	2
+	channels_min	2
+	channels_max	2
+	rate		48000
+	rate_min	16000
+	rate_max	48000
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf]: sure, updated

**Comment by RanderWang** [comment]
updated, thanks

File: tools/topology/topology2/include/components/cavs/src.conf [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+		unique "instance"
+	}
+
+	#
+	# Default attributes
+	#
+	type		"src"
+	# E61BB28D-149A-4C1F-B709-46823EF5F5AE
+	uuid 		"8d:b2:1b:e6:9a:14:1f:4c:b7:09:46:82:3e:f5:f5:ae"
```
Comment (lgirdwood) [inline tools/topology/topology2/include/components/cavs/src.conf:92]: UUID does not match the comment.

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+			Object.Base.audio_format.5 {
+				in_rate                 44100
+				in_bit_depth            32
+				in_valid_bit_depth      24
+				out_bit_depth           32
+				out_valid_bit_depth     32
+				dma_buffer_size "$[$obs * 2]"
+			}
+
+			Object.Base.audio_format.6{
```
Comment (lgirdwood) [inline tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:93]: missing space before {

File: tools/topology/topology2/include/common/audio_format.conf [hunk 6]
```diff
@@ -232,6 +232,7 @@ Class.Base."audio_format" {
 	out_fmt_cfg	"$[($out_channels | ($out_valid_bit_depth * 256)) | ($out_sample_type * 65536)]"
 
 	# math expression for computing input/put buffer sizes
-	ibs "$[($in_channels * ($in_rate / 1000)) * ($in_bit_depth / 8)]"
-	obs "$[($out_channels * ($out_rate / 1000)) * ($out_bit_depth / 8)]"
+	# for 22.05, 44.1, 88.2 and 176.4khz, we need to round it to ceiling value
+	ibs "$[($in_channels * ($[($in_rate + 950)] / 1000)) * ($in_bit_depth / 8)]"
+	obs "$[($out_channels * ($[($out_rate + 950)] / 1000)) * ($out_bit_depth / 8)]"
```
Comment (singalsu) [inline tools/topology/topology2/include/common/audio_format.conf:237]: Add of 950 is not enough to round up 11.025 kHz. + 999 should work if the arithmetic works like I'm guessing.

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	direction	"playback"
+	dynamic_pipeline 1
+	time_domain	"timer"
+	channels	2
+	channels_min	2
+	channels_max	2
+	rate		48000
+	rate_min	16000
```
Comment (singalsu) [inline tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:184]: Smallest rate should likely be 8 kHz.

File: tools/topology/topology2/include/components/gain.conf [hunk 8]
```diff
@@ -154,5 +154,4 @@ Class.Widget."gain" {
 	curve_type		"linear"
 	curve_duration		10
 	init_value		0x7fffffff
-	num_audio_formats 1
```
Comment (singalsu) [inline tools/topology/topology2/include/components/gain.conf:157]: Does this impact SRC. Should this be a separate PR?

File: tools/topology/topology2/include/components/cavs/src.conf [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+# A generic src component. All attributes defined herein are namespaced
+# by alsatplg to "Object.Widget.src.N.attribute_name"
+#
+# Usage: this component can be used by instantiating it in the parent object. i.e.
+#
+# 	Object.Widget.src."N" {
+#		period_sink_count	2
+#		period_source_count	2
+#		format			"s24le"
+#		rate_out		48000
```
Comment (plbossart) [inline tools/topology/topology2/include/components/cavs/src.conf:12]: can you describe input rates? is 48k the only supported option? what about s32le?

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+# Where N is the unique pipeline ID within the same alsaconf node.
+#
+
+<include/common/audio_format.conf>
+<include/components/copier.conf>
+<include/components/gain.conf>
+<include/components/cavs/src.conf>
+<include/components/pipeline.conf>
+
+Class.Pipeline."src-gain-playback" {
```
Comment (plbossart) [inline tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:26]: kind reminder that we want all host pipelines to go into a mixer, and a different pipeline for dais.

I don't want yet another case where the mixer is ignored. 

To be clearer, multiple playback streams is a requirement.

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf (hunk 7 above)
Comment (plbossart) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:184]: and it's inconsistent to gave rate min and max, but only describe formats for the 44.1 kHz case. Either we support all intermediate values or we don't.

File: tools/topology/topology2/include/components/cavs/src.conf (hunk 4 above)
Comment (fredoh9) [reply tools/topology/topology2/include/components/cavs/src.conf:92]: I see two SRC modules, are we adding SRCINTC? Then the UUID seems right.

```
[   61.491741] sof-audio-pci-intel-mtl 0000:00:1f.3: module SRCINTC: UUID E61BB28D-149A-4C1F-B709-46823EF5F5AE cfg_count: 23, bss_size: 0x3220                     
[   61.491745] sof-audio-pci-intel-mtl 0000:00:1f.3: module SRCLITE: UUID 33441051-44CD-466A-83A3-178478708AEA cfg_count: 8, bss_size: 0x2000
```

Comment (RanderWang) [reply tools/topology/topology2/include/components/cavs/src.conf:92]: the order is a little different but it is right

File: tools/topology/topology2/include/common/audio_format.conf (hunk 6 above)
Comment (RanderWang) [reply tools/topology/topology2/include/common/audio_format.conf:237]: o, so we can support 11025

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf (hunk 7 above)
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:184]: sure

File: tools/topology/topology2/include/components/gain.conf (hunk 8 above)
Comment (RanderWang) [reply tools/topology/topology2/include/components/gain.conf:157]: it is a issue find by src.

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+		route."2" {
+			source  "src..1"
+			sink    "gain..1"
+		}
+	}
+
+	direction	"playback"
+	dynamic_pipeline 1
+	time_domain	"timer"
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:179]: **Good question. It is cpc which is stored in module config in fw manifest, please check the CPC in the following table.  Driver should match the entry by pcm parameters and get CPC.  The total CPC value in all module will be added and transformed to cpc for pipeline creation (Maybe drive doesn't support it now). The close source fw will use this cpc to schedule the pipeline. Currently it is not used by SOF IPC3 & IPC4. May be it will be supported by IPC4 in future.**


````
	# mod_cfg [PAR_0 PAR_1 PAR_2 PAR_3 IS_BYTES CPS IBS OBS MOD_FLAGS CPC OBLS]
	mod_cfg = [0,    0,     0,    0,     12832,  1365500, 0, 0, 0,   1365, 0,
		   1, 0, 0, 0, 12832, 2302300, 0, 0, 0, 2302, 0,
		2, 0, 0, 0, 12832, 3218200, 0, 0, 0, 3218, 0,
		3, 0, 0, 0, 12832, 4169700, 0, 0, 0, 4169, 0,
		4, 0, 0, 0, 12832, 5095100, 0, 0, 0, 5095, 0,
		5, 0, 0, 0, 12832, 6014800, 0, 0, 0, 6014, 0,
		6, 0, 0, 0, 12832, 6963500, 0, 0, 0, 6963, 0,
		7, 0, 0, 0, 12832, 7791000, 0, 0, 0, 7791, 0,
		8, 0, 0, 0, 12832, 8843000, 0, 0, 0, 8843, 0,
		9, 0, 0, 0, 12832, 9755100, 0, 0, 0, 9755, 0,
		10, 0, 0, 0, 12832, 10726500, 0, 0, 0, 10726, 0,
		11, 0, 0, 0, 12832, 11624100, 0, 0, 0, 11624, 0,
		12, 0, 0, 0, 12832, 12518700, 0, 0, 0, 12518, 0,
		13, 0, 0, 0, 12832, 13555000, 0, 0, 0, 13555, 0,
		14, 0, 0, 0, 12832, 14144500, 0, 0, 0, 14144, 0,
		15, 0, 0, 0, 12832, 15809800, 0, 0, 0, 15809, 0,
		16, 0, 0, 0, 12832, 16749000, 0, 0, 0, 16749, 0,
		17, 0, 0, 0, 12832, 18433500, 0, 0, 0, 18433, 0,
		18, 0, 0, 0, 12832, 19425900, 0, 0, 0, 19425, 0,
		19, 0, 0, 0, 12832, 20396900, 0, 0, 0, 20396, 0,
		20, 0, 0, 0, 12832, 20881000, 0, 0, 0, 20881, 0,
		21, 0, 0, 0, 12832, 23431000, 0, 0, 0, 23431, 0,
		22, 0, 0, 0, 12832, 30471000, 0, 0, 0, 30471, 0]
````

**Comment by RanderWang** [comment]
change it to RFC since we need to support
(1) all the formats. The debugging and validation effort will be significantly increased
(2) add mixin & mixout

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf (hunk 10 above)
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:26]: sure, I am working on it

File: tools/topology/topology2/include/components/cavs/src.conf (hunk 9 above)
Comment (RanderWang) [reply tools/topology/topology2/include/components/cavs/src.conf:12]: I will update PR to include all supported input rate

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf (hunk 11 above)
[the remark this answers is not in the document]
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:179]: How to match it ? it is a puzzle to us. Driver uses a complex algorithm and we need time to decode it. I can't manage to understand it at first glance. Do we need to support it ? As these numbers are built with close source fw so they are not useful for sof fw in future.

**Comment by RanderWang** [comment]
@plbossart  @lgirdwood  @ranj063  @bardliao I have an idea bout SRC support in Linux kernel.
SOF can support many sample rates but it is not optimized to build it in topology. The reason is that 48khz audio file will be used in 4x% chance,  4x% 44.1khz and n% for other sample rates. If SRC is built in topology it will introduce more cpu usage for normal 48khz audio. It is not a good deal.

My solution is dynamically add SRC in driver according to sample rate so that we have optimized performance.  Another way is to delete it in driver for hardware supported rate.

File: tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf (hunk 11 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-playback.conf:179]: Some of these values are used by the Windows driver. There may be future schedulers that could use some of this data, but tooling to generate this data is also needed. Currently not needed so can be set to 0 for compatibility.

**Comment by lgirdwood** [comment]
> @plbossart @lgirdwood @ranj063 @bardliao I have an idea bout SRC support in Linux kernel. SOF can support many sample rates but it is not optimized to build it in topology. The reason is that 48khz audio file will be used in 4x% chance, 4x% 44.1khz and n% for other sample rates. If SRC is built in topology it will introduce more cpu usage for normal 48khz audio. It is not a good deal.
> 
> My solution is dynamically add SRC in driver according to sample rate so that we have optimized performance. Another way is to build it in topology but delete it in driver for no conversion rate.

@RanderWang just a thought, rather than expressing all the supported in/out rates for the SRC in the topology, we could use a wildcard rate (like SNDRV_PCM_RATE_CONTINUOUS using 0) which means we can define a single SRC object with wildcard in/out rates where the FW could reject any invalid combination.

**Comment by plbossart** [comment]
> @plbossart @lgirdwood @ranj063 @bardliao I have an idea bout SRC support in Linux kernel. SOF can support many sample rates but it is not optimized to build it in topology. The reason is that 48khz audio file will be used in 4x% chance, 4x% 44.1khz and n% for other sample rates. If SRC is built in topology it will introduce more cpu usage for normal 48khz audio. It is not a good deal.
> 
> My solution is dynamically add SRC in driver according to sample rate so that we have optimized performance. Another way is to build it in topology but delete it in driver for no conversion rate.

That's what others have suggested in the 'AVS' driver, you could bind modules only if required, e.g. if you don't need an instance of SRC you can bypass the SRC.

The problem is that it's of course a lot more complicated to decide whether to bind or not. This would not work for e.g. multiple steps of conversions, because at the moment the DPCM driver cannot be informed of possible rate changes.

Also the introduction of the SRC is a nice feature, but who needs it? In practice, sound servers (CRAS, PulseAudio, etc) already resample to a desired rate, so it's not like we can expect tons of different configurations. Specifically the 44.1kHz case is not required, we've lived since 2013 without SRC in the firmware and no one ever complained about the host-based conversion.

And last, SRC takes resources that could be used for other things that have more added value, such as branded processing, speaker amp. SRC is an item on a program manager checklist, its value as part of the firmware is very very questionable with very very limited returned IMHO. I vote that we don't spend too much time on the optimization and just focus on getting SRC to work. It remains to be seen whether this will actually be used in a product.

**Comment by RanderWang** [comment]
updated.
(1) support all the rate from 8k to 192k
(2) add mixer support

Validated on ADL

File: tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	}
+
+	Object.Widget {
+		copier."1" {
+			copier_type	"host"
+			type	"aif_in"
+			num_audio_formats 3
+
+			# to avoid create dozens of format entrys use
+			# SNDR_PCM_RATE_8000_192000 which is 0x1FFE
```
Comment (lgirdwood) [inline tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf:56]: Can we define these as macros somewhere so we can use them instead of 0x1FFE ?

File: tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+#
+# List of all DAIs
+#
+#ALH Index: 0, Direction: duplex
+Object.Dai {
+	ALH."2" {
+		id 		2
+		direction	"playback"
+		name		SDW2-Playback
```
Comment (fredoh9) [inline tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf:45]: SDW2-Playback and SDW3-Capture, are this device specific? Which device are you testing with?

File: tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+				out_rate                0x1FFE
+				out_bit_depth		32
+				out_valid_bit_depth	24
+				dma_buffer_size "$[$obs * 2]"
+			}
+
+			Object.Base.audio_format.2 {
+                                in_rate                 0x1FFE
+                                out_rate                0x1FFE
+				in_bit_depth		32
```
Comment (fredoh9) [inline tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf:69]: 16bit bit depth is not consideration? What if input file has 16bits per sample? will fail?

File: tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf (hunk 13 above)
Comment (RanderWang) [reply tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf:45]: ANDREWS

File: tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf (hunk 14 above)
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf:69]: above one is for 16bits

File: tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf (hunk 13 above)
Comment (RanderWang) [reply tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf:45]: Adjust it

File: tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf (hunk 12 above)
Comment (RanderWang) [reply tools/topology/topology2/include/pipelines/cavs/src-gain-mixin-playback.conf:56]: sure, updated

**Comment by RanderWang** [comment]
updated, thanks!

**Review by singalsu — APPROVED** [review]
Looks OK to me for discussed part: frames count calculation.

File: tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf (hunk 13 above)
Comment (fredoh9) [reply tools/topology/topology2/cavs/cavs-sdw-src-gain-mixin.conf]: I know we need to change this and doing it. My point was cavs-sdw.conf has SDW0-Playback and SDW0-Capture, we should match this for consistency.

**Review by lgirdwood — APPROVED** [review]

**Comment by XiaoyunWu6666** [comment]
Hi @RanderWang , this PR seem to cause some error in cavs-mixin-mixout-hda.tplg and cause multiple IPC4 HDA platforms failed to load the tplg correctly.
Do you have any idea? Thanks

**Comment by bardliao** [comment]
> Hi @RanderWang , this PR seem to cause some error in cavs-mixin-mixout-hda.tplg and cause multiple IPC4 HDA platforms failed to load the tplg correctly. Do you have any idea? Thanks

@XiaoyunWu6666 Please try https://github.com/thesofproject/sof/pull/5971
